src/memcached/cache_impl.go

```diff
@@ -0,0 +1,294 @@
+// The memcached limiter uses GetMulti() to check keys in parallel and then does
+// increments asynchronously in the backend, since the memcache interface doesn't
+// support multi-increment and it seems worthwhile to minimize the number of
+// concurrent or sequential RPCs in the critical path.
+//
+// Another difference from redis is that memcache doesn't create a key implicitly by
+// incrementing a missing entry. Instead, when increment fails an explicit "add" needs
+// to be called. The process of increment becomes a bit of a dance since we try to
+// limit the number of RPCs. First we call increment, then add if the increment
+// failed, then increment again if the add failed (which could happen if there was
+// a race to call "add").
+
+package memcached
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"strconv"
+	"sync"
+
+	"github.com/coocood/freecache"
+	stats "github.com/lyft/gostats"
+
+	"github.com/bradfitz/gomemcache/memcache"
+
+	logger "github.com/sirupsen/logrus"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+
+	"github.com/envoyproxy/ratelimit/src/assert"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/settings"
+	"github.com/envoyproxy/ratelimit/src/utils"
+)
+
+type rateLimitMemcacheImpl struct {
+	client                     Client
+	timeSource                 utils.TimeSource
+	jitterRand                 *rand.Rand
+	expirationJitterMaxSeconds int64
+	cacheKeyGenerator          limiter.CacheKeyGenerator
+	localCache                 *freecache.Cache
+	wg                         sync.WaitGroup
+	nearLimitRatio             float32
+}
+
+var _ limiter.RateLimitCache = (*rateLimitMemcacheImpl)(nil)
+
+func max(a uint32, b uint32) uint32 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func (this *rateLimitMemcacheImpl) DoLimit(
+	ctx context.Context,
+	request *pb.RateLimitRequest,
+	limits []*config.RateLimit) []*pb.RateLimitResponse_DescriptorStatus {
+
+	logger.Debugf("starting cache lookup")
+
+	// request.HitsAddend could be 0 (default value) if not specified by the caller in the Ratelimit request.
+	hitsAddend := max(1, request.HitsAddend)
```

Comment:
I see that bunch of code is duplicated across redis and memcache implementations particularly in this method. Would it be possible to factor that code our into some common method?

Reply:
Note this comment is the same as this one: https://github.com/envoyproxy/ratelimit/pull/193#discussion_r532778253. It would be great to figure out some of this as I think the two PRs are going to run into very similar issues around duplication so I would take a look at my other comments in that one also.

Reply:
ok, maybe we need a separate PR which pulls up duplicated code into single place